lib/ansible/modules/crypto/openssh_keypair.py

```diff
@@ -0,0 +1,327 @@
+    type: dsa
+'''
+
+RETURN = '''
+size:
+    description: Size (in bits) of the SSH private key
+    returned: changed or success
+    type: int
+    sample: 4096
+type:
+    description: Algorithm used to generate the SSH private key
+    returned: changed or success
+    type: string
+    sample: rsa
+filename:
+    description: Path to the generated SSH private key file
+    returned: changed or success
+    type: string
+    sample: /tmp/id_ssh_rsa
+fingerprint:
+    description: The fingerprint of the key.
+    returned: changed or success
+    type: string
+    sample: 4096 SHA256:r4YCZxihVjedH2OlfjVGI6Y5xAYtdCwk8VxKyzVyYfM example@example.com (RSA)
+'''
+
+import os
+import errno
+import subprocess
+
+from subprocess import PIPE, Popen
+from ansible.module_utils.basic import AnsibleModule
+from ansible.module_utils._text import to_native
+
+
+class PrivateKeyError(Exception):
+    pass
+
+
+class PrivateKey(object):
+
+    def __init__(self, module):
+        self.path = module.params['path']
+        self.state = module.params['state']
+        self.force = module.params['force']
+        self.size = module.params['size']
+        self.type = module.params['type']
+        self.passphrase = module.params['passphrase']
+        self.comment = module.params['comment']
+        self.changed = False
+        self.check_mode = module.check_mode
+        self.privatekey = None
+        self.fingerprint = {}
+
+        if self.type in ('rsa', 'rsa1'):
+            self.size = 4096 if self.size is None else self.size
+            if self.size < 1024:
+                module.fail_json(msg='''For RSA keys, the minimum size is 1024 bits and the default is 4096 bits.
+                                     Attempting to use bit lengths under 1024 will cause the module to fail.''')
+
+        if self.type == 'dsa':
+            self.size = 1024 if self.size is None else self.size
+            if self.size != 1024:
+                module.fail_json(msg='''DSA keys must be exactly 1024 bits as specified by FIPS 186-2.''')
+
+        if self.type == 'ecdsa':
+            self.size = 256 if self.size is None else self.size
+            if self.size not in (256, 384, 521):
+                module.fail_json(msg='''For ECDSA keys, size determines the key length by selecting from
+                                     one of three elliptic curve sizes: 256, 384 or 521 bits.
+                                     Attempting to use bit lengths other than these three values for
+                                     ECDSA keys will cause this module to fail.''')
+        if self.type == 'ed25519':
+            self.size = 256
+
+    def generate(self, module):
+        # generate a keypair
+        if not self.check(module, perms_required=False) or self.force:
+            args = [
+                'ssh-keygen',
+                '-q',
+                '-b', str(self.size),
+                '-t', self.type,
+                '-f', self.path,
+            ]
```

Comment:
It should be possible to do this in Python I think... But that may be more complicated. I know PyCrypto can create RSA and P256 keys, but I don't think it supports ed25519. If you choose to stick to calling `ssh-keygen`, that's fine, but just so you're aware.

Reply:
I would stick to [cryptography](https://cryptography.io/en/latest/). PyCrypto was last updated in 2013 (!), so it seems pretty dead.

Reply:
The same issue was already happening with the `openssl_*` modules (pyOpenSSL is EOL already for a while and was even already mostly abandoned when these modules were written) - it is extremely hard to support even the most recent LTS distributions. The goal for the `openssl_*` ones that we needed internally was Debian Jessie (with backports being acceptable) --> https://packages.debian.org/jessie-backports/python-openssl

This is even an issue even on RHEL/CentOS latest (https://pkgs.org/download/pyopenssl), where some parts of the `openssl_*` modules don't work.

Potentially using/requiring `cryptography` 1.7 might be an option for this one, but writing parts of openssh-keygen in Python (probably Python 2+3 by the way for the next few years) is also not a great thing to do, but doesn't seem to hard for simple key files: https://stackoverflow.com/a/39126754

This might get hairy though if more advanced features get used.

Reply:
By the way, we plan to also write and ship a module for OpenSSH certificates (https://blog.habets.se/2011/07/OpenSSH-certificates.html) and these are even worse concerning support in any established library. That means for that one we need `ssh-keygen` anyways, so having this module use it too is more consistent. I'm sure that patches would be welcome though if someone wants to rewrite this module to use `cryptography` (maybe optionally?), shouldn't be too hard to do.